tsdb/querier.go

```diff
@@ -326,50 +326,39 @@ func findSetMatches(pattern string) []string {
 // PostingsForMatchers assembles a single postings iterator against the index reader
 // based on the given matchers.
 func PostingsForMatchers(ix IndexReader, ms ...labels.Matcher) (index.Postings, error) {
+	var err error
+	var it index.Postings
 	var its, notIts []index.Postings
-	// See which label must be non-empty.
-	labelMustBeSet := make(map[string]bool, len(ms))
-	for _, m := range ms {
-		if !m.Matches("") {
-			labelMustBeSet[m.Name()] = true
-		}
-	}
 
 	for _, m := range ms {
 		matchesEmpty := m.Matches("")
-		if labelMustBeSet[m.Name()] || !matchesEmpty {
+		nm, isNot := m.(*labels.NotMatcher)
+		if !matchesEmpty {
 			// If this matcher must be non-empty, we can be smarter.
-			nm, isNot := m.(*labels.NotMatcher)
-			if isNot && matchesEmpty { // l!="foo"
-				// If the label can't be empty and is a Not and the inner matcher
-				// doesn't match empty, then subtract it out at the end.
-				it, err := postingsForMatcher(ix, nm.Matcher)
-				if err != nil {
-					return nil, err
-				}
-				notIts = append(notIts, it)
-			} else if isNot && !matchesEmpty { // l!=""
+			if isNot { // l!=""
 				// If the label can't be empty and is a Not, but the inner matcher can
 				// be empty we need to use inversePostingsForMatcher.
-				it, err := inversePostingsForMatcher(ix, nm.Matcher)
-				if err != nil {
-					return nil, err
-				}
-				its = append(its, it)
+				it, err = inversePostingsForMatcher(ix, nm.Matcher)
 			} else { // l="a"
 				// Non-Not matcher, use normal postingsForMatcher.
-				it, err := postingsForMatcher(ix, m)
-				if err != nil {
-					return nil, err
-				}
-				its = append(its, it)
+				it, err = postingsForMatcher(ix, m)
 			}
-		} else { // l=""
+			if err != nil {
+				return nil, err
+			}
+			its = append(its, it)
+		} else { // l!="foo" or l=""
+			// If the label can be empty, subtract all not matched series out.
```

Comment:
These comments can be put individually inside the corresponding `if` and `else` blocks.